Implement action key path to implicit declaration

Den-Rimus opened the pull request:
Support multiple key params
Rename annotation
Adjust sample events and tests

Commit:
Implement action key path to implicit declaration

Support multiple key params
Rename annotation
Adjust sample events and tests

almozavr commented on service-compiler/src/main/java/io/techery/analytics/compiler/ActionClassUtils.java:
```diff
@@ -89,6 +91,32 @@ public static String getAnnotatedFieldName(Elements elementUtils, TypeElement ty
 
+   public static Set<KeyPathEntity> getKeyPathEntities(Elements elementUtils, TypeElement typeElement) {
+      final Set<KeyPathEntity> keyPathEntities = new HashSet<KeyPathEntity>();
+      final String keyPathSymbol = "$"; // TODO make '$' customizable
```
should be `const` to be easily identifiable durink quick class look through

almozavr commented on service-compiler/src/main/java/io/techery/analytics/compiler/ActionClassUtils.java:
```diff
@@ -89,6 +91,32 @@ public static String getAnnotatedFieldName(Elements elementUtils, TypeElement ty
+   }
+
+   public static boolean checkHasKeyParams(String actionKey) {
+      final String keyParamSymbol = "$"; // TODO: make this customizable via gradle
```
same const

almozavr commented on service-compiler/src/main/java/io/techery/analytics/compiler/ActionHelperGenerator.java:
```diff
@@ -79,8 +80,15 @@ private CodeBlock makeGetActionCodeBlock(AnalyticActionClass actionClass) {
-         builder.addStatement("return $S.replace($S, action.$L)",
-               actionClass.action, ACTION_PATH_PARAM, actionClass.actionPartFieldName);
+         builder.addStatement("$T $L = $S", String.class, GENERATED_ACTION_KEY_VAR_NAME, actionClass.action);
+         for (KeyPathEntity keyPathEntity : actionClass.keyPathEntities) {
```
in terms of performance, it's no so good to have multiple lines of string instances with replace method -> so better to replace actual string with it's parts and `+` signs

Den-Rimus replied:
Looking at effort needed to implement this I'd say that it does not worth it.
Please approve that this is really a blocker

almozavr replied:
it's not a blocker, an improvement

almozavr commented on service-compiler/src/main/java/io/techery/analytics/compiler/model/AnalyticActionClass.java:
```diff
@@ -51,11 +50,10 @@ private void obtainAttributes(Elements elementUtils, TypeElement typeElement) {
-            if (ActionClassUtils.checkHasAnnotatedField(elementUtils, typeElement, ActionPart.class)) {
-               actionPartFieldName = ActionClassUtils.getAnnotatedFieldName(elementUtils, typeElement, ActionPart.class);
-            }
+         if (ActionClassUtils.checkHasAnnotatedField(elementUtils, typeElement, KeyPath.class)) {
```
there should be validation or TODO for validation: if there are more pathes then `KeyPath` fields

almozavr reviewed (changes requested):
Also I'd like to discuss replaces logic: do we allow multiple same-name key pathes? If no (I suppose) we should validate such case.

Commit:
Code-review fixes

Add TODO's
Add small improvement for helper generated code (redundant variable creation)
Swap variables usage with constant
